File: backend/models/exam_package.go
package models

import (
	"time"
)

// ExamPackage adalah struktur file .exam yang didownload siswa
type ExamPackage struct {
	Header  ExamHeader  `json:"header"`
	Payload string      `json:"payload"` // Encrypted JSON string of ExamPayload
	Signature string    `json:"signature"` // HMAC/RSA signature of Header+Payload
}

// ExamHeader berisi metadata publik (tidak terenkripsi) agar app bisa validasi awal
type ExamHeader struct {
	ExamID        string    `json:"exam_id" gorm:"primaryKey"`
	Title         string    `json:"title"`
	Version       int       `json:"version"`
	EncryptionIV  string    `json:"iv"` // Initialization Vector untuk AES
	ValidFrom     time.Time `json:"valid_from"`
	ValidUntil    time.Time `json:"valid_until"`
	DurationMins  int       `json:"duration_mins"`
	CreatedAt     time.Time `json:"created_at"`
	
	// Relation
	Questions []Question `json:"-" gorm:"foreignKey:ExamID"`
}

// ExamPayload adalah isi soal yang dienkripsi
type ExamPayload struct {
	Questions []Question `json:"questions"`
	Config    ExamConfig `json:"config"`
}

type ExamConfig struct {
	AllowWifi       bool `json:"allow_wifi"`
	RandomizeOrder  bool `json:"randomize_order"`
	ShowResult      bool `json:"show_result"`
}

type Question struct {
	ID      string   `json:"id" gorm:"primaryKey"`
	ExamID  string   `json:"exam_id"` // FK
	Type    string   `json:"type"` // "mc", "essay"
	Content string   `json:"content"` // HTML/Markdown
	Options []Option `json:"options,omitempty" gorm:"foreignKey:QuestionID"`
	Points  int      `json:"points"`
}

type Option struct {
	DBID       uint   `json:"-" gorm:"primaryKey"`
	QuestionID string `json:"-"`
	
	ID         string `json:"id" gorm:"column:option_label"` // "A", "B"
	Content    string `json:"content"`
	IsCorrect  bool   `json:"is_correct,omitempty"` 
}

File: mobile/lib/models/exam_package.dart
// models/exam_package.dart

class ExamPackage {
  final ExamHeader header;
  final String payload; // Encrypted string
  final String signature;

  ExamPackage({required this.header, required this.payload, required this.signature});

  factory ExamPackage.fromJson(Map<String, dynamic> json) {
    return ExamPackage(
      header: ExamHeader.fromJson(json['header']),
      payload: json['payload'],
      signature: json['signature'],
    );
  }

  Map<String, dynamic> toJson() => {
    'header': header.toJson(),
    'payload': payload,
    'signature': signature,
  };
}

class ExamHeader {
  final String examId;
  final String title;
  final int version;
  final String encryptionIV;
  final DateTime validFrom;
  final DateTime validUntil;
  final int durationMins;

  ExamHeader({
    required this.examId,
    required this.title,
    required this.version,
    required this.encryptionIV,
    required this.validFrom,
    required this.validUntil,
    required this.durationMins,
  });

  factory ExamHeader.fromJson(Map<String, dynamic> json) {
    return ExamHeader(
      examId: json['exam_id'],
      title: json['title'],
      version: json['version'],
      encryptionIV: json['iv'],
      validFrom: DateTime.parse(json['valid_from']),
      validUntil: DateTime.parse(json['valid_until']),
      durationMins: json['duration_mins'],
    );
  }

  Map<String, dynamic> toJson() => {
    'exam_id': examId,
    'title': title,
    'version': version,
    'iv': encryptionIV,
    'valid_from': validFrom.toIso8601String(),
    'valid_until': validUntil.toIso8601String(),
    'duration_mins': durationMins,
  };
}

// Decrypted Payload
class ExamPayload {
  final List<Question> questions;
  final ExamConfig config;

  ExamPayload({required this.questions, required this.config});

  factory ExamPayload.fromJson(Map<String, dynamic> json) {
    var list = json['questions'] as List;
    List<Question> questionsList = list.map((i) => Question.fromJson(i)).toList();
    return ExamPayload(
      questions: questionsList,
      config: ExamConfig.fromJson(json['config']),
    );
  }
}

class ExamConfig {
  final bool allowWifi;
  final bool randomizeOrder;

  ExamConfig({required this.allowWifi, required this.randomizeOrder});

  factory ExamConfig.fromJson(Map<String, dynamic> json) {
    return ExamConfig(
      allowWifi: json['allow_wifi'] ?? false,
      randomizeOrder: json['randomize_order'] ?? false,
    );
  }
}

class Question {
  final String id;
  final String type;
  final String content;
  final List<Option> options;
  final int points;

  Question({required this.id, required this.type, required this.content, required this.options, required this.points});

  factory Question.fromJson(Map<String, dynamic> json) {
    var opts = json['options'] != null ? (json['options'] as List).map((i) => Option.fromJson(i)).toList() : <Option>[];
    return Question(
      id: json['id'],
      type: json['type'],
      content: json['content'],
      options: opts,
      points: json['points'],
    );
  }
}

class Option {
  final String id;
  final String content;

  Option({required this.id, required this.content});

  factory Option.fromJson(Map<String, dynamic> json) {
    return Option(id: json['id'], content: json['content']);
  }
}
